File: sliding-window/longest-substring-without-repeating.py
# return length of longest substring
def find_longest_substring(string):
    last_char_index = {}
    start = 0
    sub_len = 0

    for end, char in enumerate(string):
        if char in last_char_index:
            start = max(start, last_char_index[char] + 1)

        sub_len = max(sub_len, end - start + 1)

        last_char_index[char] = end

    return sub_len

File: sliding-window/longest-substring-without-repeating.js
export function findLongestSubstring(str) {
	let lastSeenAt = {};
	let start = 0;
	let longest = 0;

	for (let end = 0; end < str.length; end++) {
		const curr = str[end];

		if (curr in lastSeenAt) {
			start = Math.max(start, lastSeenAt[curr] + 1);
		}

		longest = Math.max(longest, end - start + 1);

		lastSeenAt[curr] = end;
	}

	return longest;
}
